### DISCUSSION TITLE
What would be the most suitable collider shape for a voxel terrain?

### DISCUSSION THREAD

**Discussion by SirLynix** [jrouwe/JoltPhysics#552] [Q&A] [answered]
Hi there,

I'm working on a game with planets made of voxels ("Minecraft-like"), and I'm wondering what would be the most efficient way to handle collisions.

Here's a (very WIP) screenshot of the voxel grid:
![image](https://github.com/jrouwe/JoltPhysics/assets/3002461/f8575617-3343-4516-9c76-006d20c670d8)

I'm thinking of simple AABB collisions, but I'm not sure Jolt has a suitable shape for this.

Also, it's not just a voxel grid, it's a cube planet with round edges:
![image](https://github.com/jrouwe/JoltPhysics/assets/3002461/8ab527e6-c10e-4e0a-899d-41bfda8a6dbe)

The grid will be made of different chunks, and only some of them would be round (curved), so I can handle this in two ways. I was thinking of using mesh shapes (aka triangle shapes) for rounded parts but wondering about the flat parts.

Of course, as with Minecraft, players are expected to be able to mine through the planet, regenerating the visual and collision meshes regularly, so shape creation time plays here.

(of course, the planet chunks are static).

Any idea here?

Thank you.

**Comment by jrouwe** [comment]
There was an earlier discussion about this: #446

**Reply by SirLynix** [reply]
Thanks! I was thinking also of custom shapes which seems to be the answer for the flat faces. However as for the curved chunks am I right to think there's no better way than a (triangle) mesh shape?

![image](https://github.com/jrouwe/JoltPhysics/assets/3002461/074927b0-0d30-4d6c-aeff-ae0cc7fae4c0)

**Reply by jrouwe** [reply]
You can still use the same approach but with a MeshShape instead of a BoxShape inside (I'm assuming there will be a finite number of 'curved chunks'?). This will have the disadvantage that there will be seams between the chunks that can cause ghost collisions, in order to avoid those you'd need to do the triangle handling yourself in your custom shape so you can specify the 'active edges'. See TriangleShape::sCollideConvexVsTriangle and sCastConvexVsTriangle as an example.

Making a custom shape is quite a bit of work, so if you have enough memory to just flatten everything to a MeshShape then I'd say that that would be the simplest solution.

**Reply by SirLynix** [reply]
I'd like to have much bigger planets than the one shown here, but you're right that there will be less curved chunks than flat ones.

Are you saying a static compound shape with a lot of box shapes be suitable for flat chunks (performance-wise)? or that a mesh shape would be better to handle both curved and flat chunks?

Also, regarding the player collider (which is currently a capsule), would there a big cost difference versus a box shape (or custom collider to make axis-aligned boxes), if I were to use a mesh shape (thinking about curved chunks here, for flat chunks made of AABB I think it's pretty clear that AABB <-> AABB collisions would be incredibly efficient).

Of course I will benchmark this, but I'm taking all advices 😄

**Reply by jrouwe** [reply, marked as answer]
> Are you saying a static compound shape with a lot of box shapes be suitable for flat chunks (performance-wise)? or that a mesh shape would be better to handle both curved and flat chunks?

If you would need 1000x1000 boxes to cover a flat area that you could otherwise cover with 2 triangles then obviously the triangles are going to win, so the answer is: it depends. In general I would say testing a convex against 1 triangle is roughly the same as testing a convex against 1 box (it's the same code).

> Also, regarding the player collider (which is currently a capsule), would there a big cost difference versus a box shape (or custom collider to make axis-aligned boxes), if I were to use a mesh shape (thinking about curved chunks here, for flat chunks made of AABB I think it's pretty clear that AABB <-> AABB collisions would be incredibly efficient).

You wouldn't be doing AABB vs AABB but capsule vs AABB which uses the general convex vs convex algorithm. So again it's roughly the same cost as testing a capsule vs 1 triangle. Obviously it's possible to specialize the collision functions to increase performance, but I've only made specific code paths for sphere vs X and not for capsule since we happen to be testing a lot less of those.

I think you really need to profile your situation and beware of any ghost collisions you introduce by composing your shape of multiple sub shapes (for a box you can reduce those by increasing the convex radius).
